### ISSUE TITLE
TypeError: Cannot set property 'request' of null

### ISSUE THREAD

**Issue by consense** [elastic/apm-agent-nodejs#235]
Versions:

- node: 8.9.4
- apm-agent-nodejs: 1.0.0
- elasticsearch, apm-server: 6.2

the nodejs client throws (briefly after start when first sample is apparently processed) with
```
TypeError: Cannot set property 'request' of null
    at /APPDIR/node_modules/elastic-apm-node/lib/instrumentation/transaction.js:160:31
    at /APPDIR/node_modules/after-all-results/index.js:9:18
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```

corresponding entry in apm-server log:
`2018-02-08T10:55:33.755Z        INFO    [request]       beater/handlers.go:150  handled request {"request_id": "c64c3d8b-e467-4f79-91ce-0af2d185a214", "response_code": 202, "method": "POST", "URL": "/v1/errors", "content_length": 1410, "remote_address": "9.8.7.6", "user_agent": "elastic-apm-node/1.0.0 elastic-apm-http-client/5.1.0"}`

content of elastic-apm-node.js:
```
module.exports = {
	active: process.env.NODE_ENV === 'production',
	serviceName: 'service-name',
	serverUrl: 'http://1.2.3.4:8200',
	flushInterval: 5,
	maxQueueSize: 40,
	stackTraceLimit: 100,
	transactionSampleRate: 0.1
}
```

gladly tell me what other information would be helpful :-)

**Comment by watson** [comment]
@consense thanks for reporting this bug. I'm sorry you had to encounter it. I've opened a PR with a fix now and we'll release 1.0.1 with it later today.

The bug was introduced as part of our new sampling logic added in version 1.0.0, and apparently our test suite wasn't testing all cases 😬

On another note, it's great to see that you're making use of the different config options for tuning the performance of the agent 😃 I'd love to know more about your reasons and experience doing so. How did you come up with the different values etc? E.g. our default `stackTraceLimit` is `50`, but is that too low in your experience?

**Comment by consense** [comment]
@watson quite the contrary - thanks for looking into this so quickly - amazing!

Reason for trying to tune the performance stem from the experience we had as mentioned in #193 - CPU/memory usage increasing massively when enabling APM. Your comment regarding stack trace # might of course be a giant pointer in this matter (think back then I saw the setting and had wrongly assumed 100 would be lower than your fdefault) - allow me to run with adjusted settings for an hour :-)

**Comment by watson** [comment]
We made a few changes in v1 that drastically reduced the memory and CPU overhead of reading the source code. I think that was our biggest CPU/memory hog. We used an extremely unoptimized way of extracting specific lines from the source files. This is now much more efficient and is in most cases performed in less than one millisecond.

And on top of this, we changed the default settings to not include source code for span stack traces. You can enable this again if you like using [`sourceLinesSpanAppFrames` and `sourceLinesSpanLibraryFrames`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#source-context-span).

In fact we changed a lot of the defaults to be more performance friendly in the new version of the agent. The new limit of `50` for [`stackTraceLimit`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#stack-trace-limit) was also introduced in version 1.0.0. The previous limit was `300` if I remember correctly.

**Comment by watson** [comment]
The defaults in version 1.x for the 4 settings you use are:

- `flushInterval`: 10 seconds
- `maxQueueSize`: None (Infinity)
- `stackTraceLimit`: 50
- `transactionSampleRate`: 1.0 (i.e. no sampling)

**Comment by consense** [comment]
thanks - reducing stackTraceLimit to 5 only doubled usual cpu usage, which still seems high but than again leaves more than enough room to actually use APM for the next time without sampling.
Will ping you in case I notice anything unusual. 
From a feature perspective both apm-server (inside kibana) as well as the agent are a joy to use - kudos to you and the rest of the elastic team!

**Comment by watson** [comment]
> reducing stackTraceLimit to 5 only doubled usual cpu usage

The CPU usage went up when lowering the `stackTraceLimit`?

**Comment by consense** [comment]
no, usual cpu usage (digital ocean standard 3gb 1cpu machine) lies at 10%, with `stackTraceLimit` 100 it peaked at 100. Using `stackTraceLimit` results in 20% which is perfectly fine for the current use case I have

**Comment by consense** [comment]
actually have to go back on yesterday's comment. we are still seeing instances where the service which operates normally without the agent enabled will - with agent enabled - run smoothly (only 50%-100% cpu usage increase compared to no agent) for like half and hour but then cpu and memory will climb with the (arguably weak) appserver getting bogged down on 100% cpu usage.

**Comment by watson** [comment]
We just released version 1.0.1 with the fix for the sampling issue you discovered.

Regarding the CPU usage: Could I get you to send me an e-mail with some more data on your application, e.g RPM when you see the CPU increase and possibly also a list of your dependencies from the package.json? Then I'd love to see if I can figure out what's going on: watson@elastic.co
